### PR TITLE
added recipe for boost-unit-definitions

### PR DESCRIPTION
**Pull request by CD3**
boost-unit-definitions/0.2.2

A small, single header, library with some useful pre-defined Boost.Units unit types and instances.

I am the author of this library and wanted something small to contribute before trying to contribute some of my larger libs.

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by uilianries** [comment]
@CD3 Thank you for your contribution. First I need to ask few things about your project, is it an official part of Boost or is it in the middle of a process to be part? If yes, we can't accept it, instead, we only give support to the Boost package, because splitting Boost modules is a nightmare to manage (past experience) and we decided to not keeping other packages from boost here.

Otherwise, in case you only named your project with Boost, but it's not part of Boost project, please, rename the package to `cd3-boost-unit-definitions` to avoid name collision and keep it separated from boost project. We discussed about it recently and your case it's documented here: https://github.com/conan-io/conan-center-index/pull/14674

Thank you for providing this recipe.

**Comment by CD3** [comment]
It is not part of boost, it is just a small convenience library on top of boost's units library. I renamed the package to cd3-boost-unit-definitions as suggested. Thank you for considering the recipe.

File: recipes/cd3-boost-unit-definitions/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.build import check_min_cppstd
+from conan.tools.files import (
+    apply_conandata_patches,
+    export_conandata_patches,
+    get,
+    copy,
+)
```
Comment (prince-chrismc) [inline recipes/cd3-boost-unit-definitions/all/conanfile.py:9]: Linter

```suggestion
from conan.tools.files import get, copy
```

File: recipes/cd3-boost-unit-definitions/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+
+required_conan_version = ">=1.52.0"
+
+
+class PackageConan(ConanFile):
+    name = "cd3-boost-unit-definitions"
+    description = "A collection of pre-defined types and unit instances for working with Boost.Units quantities."
+    # Use short name only, conform to SPDX License List: https://spdx.org/licenses/
+    # In case not listed there, use "LicenseRef-<license-file-name>"
```
Comment (prince-chrismc) [inline recipes/cd3-boost-unit-definitions/all/conanfile.py:22]: Please clean up some of the comments from the template 🙏

File: recipes/cd3-boost-unit-definitions/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    # no exports_sources attribute, but export_sources(self) method instead
+    # this allows finer grain exportation of patches per version
+    def export_sources(self):
+        # export_conandata_patches(self)
+        pass
+
+    def layout(self):
+        # src_folder must use the same source folder name the project
+        cmake_layout(self)
```
Comment (prince-chrismc) [inline recipes/cd3-boost-unit-definitions/all/conanfile.py:55]: As this is header only and not using the CMake helpers we can use the `basic_layout`

```suggestion
        basic_layout(self)
```

File: recipes/cd3-boost-unit-definitions/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            src=os.path.join(self.source_folder, "src"),
+        )
+
+    def package_info(self):
+        # folders not used for header-only
+        self.cpp_info.bindirs = []
+        self.cpp_info.libdirs = []
+
+        # if package provides a CMake config file (package-config.cmake or packageConfig.cmake, with package::package target, usually installed in <prefix>/lib/cmake/<package>/)
+        self.cpp_info.set_property("cmake_file_name", "BoostUnitDefinitions")
```
Comment (prince-chrismc) [inline recipes/cd3-boost-unit-definitions/all/conanfile.py:112]: 👍 https://github.com/CD3/BoostUnitDefinitions/blob/0e0b14255644a1de16d7a3a02fcd00018ecabaa4/CMakeLists.txt#L57

**Review by prince-chrismc — COMMENTED** [review]
Overall this looks really good, just a few minor refinements.

I would appreciate if you could remove some of the template comments, they make a lot of noise for review and maintenance 🙏

File: recipes/cd3-boost-unit-definitions/all/conanfile.py (hunk 4 above)
Comment (CD3) [reply recipes/cd3-boost-unit-definitions/all/conanfile.py:107]: I'm not sure what this one means. Should I put a link to the BoostUnitDefinitions project in the recipe here as a reference for the cmake file name?

File: recipes/cd3-boost-unit-definitions/all/conanfile.py (hunk 2 above)
Comment (CD3) [reply recipes/cd3-boost-unit-definitions/all/conanfile.py:17]: Sorry, I'll clean them all out.

File: recipes/cd3-boost-unit-definitions/all/conanfile.py (hunk 4 above)
Comment (prince-chrismc) [reply recipes/cd3-boost-unit-definitions/all/conanfile.py:87]: Oh sorry, I just left a comment for my future self so I know I looked this up the first time 😉 

No action is required

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

File: recipes/cd3-boost-unit-definitions/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+class PackageConan(ConanFile):
+    name = "cd3-boost-unit-definitions"
+    description = "A collection of pre-defined types and unit instances for working with Boost.Units quantities."
+    license = "MIT"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/CD3/BoostUnitDefinitions"
+    topics = ("physical dimensions", "header-only")
+    settings = "os", "arch", "compiler", "build_type"  # even for header only
+    no_copy_source = True  # do not copy sources to build folder for header only projects, unless, need to apply patches
```
Comment (SpaceIm) [inline recipes/cd3-boost-unit-definitions/all/conanfile.py:21]: comments are not needed

File: recipes/cd3-boost-unit-definitions/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+    @property
+    def _min_cppstd(self):
+        return 14
+
+    @property
+    def _compilers_minimum_version(self):
+        return {
+            "Visual Studio": "15",
+            "msvc": "14.1",
```
Comment (SpaceIm) [inline recipes/cd3-boost-unit-definitions/all/conanfile.py:31]: it's 191 for msvc version

File: recipes/cd3-boost-unit-definitions/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+    def export_sources(self):
+        pass
+
+    def layout(self):
+        basic_layout(self)
+
+    def requirements(self):
+        self.requires("boost/1.72.0", transitive_headers=True)
```
Comment (SpaceIm) [inline recipes/cd3-boost-unit-definitions/all/conanfile.py:44]: why such old version?

File: recipes/cd3-boost-unit-definitions/all/test_package/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.8)
+project(test_package LANGUAGES CXX) # if the project uses c++
+
+find_package(BoostUnitDefinitions REQUIRED CONFIG)
+
+add_executable(${PROJECT_NAME} test_package.cpp)
+# don't link to ${CONAN_LIBS} or CONAN_PKG::package
+target_link_libraries(${PROJECT_NAME} PRIVATE BoostUnitDefinitions::BoostUnitDefinitions)
+# In case the target project need a specific C++ standard
```
Comment (SpaceIm) [inline recipes/cd3-boost-unit-definitions/all/test_package/CMakeLists.txt:9]: all these comments should be removed

File: recipes/cd3-boost-unit-definitions/all/conanfile.py (hunk 7 above)
Comment (CD3) [reply recipes/cd3-boost-unit-definitions/all/conanfile.py:44]: Good question. It's just the version I have been using in my projects. It really does not require a specific version, any of the version in conancenter will work. Should I make this a version range?

File: recipes/cd3-boost-unit-definitions/all/test_package/CMakeLists.txt (hunk 8 above)
Comment (CD3) [reply recipes/cd3-boost-unit-definitions/all/test_package/CMakeLists.txt:9]: removed

File: recipes/cd3-boost-unit-definitions/all/conanfile.py (hunk 7 above)
Comment (CD3) [reply recipes/cd3-boost-unit-definitions/all/conanfile.py:44]: Updated this to
```
self.requires("boost/[>=1.69.0]", transitive_headers=True)
```

**Pull request merged**